pfhedge/nn/functional.py

```diff
@@ -321,6 +322,75 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
     return output
 
 
+def quadratic_cvar(input: Tensor, lam: float, dim: Optional[int] = None) -> Tensor:
+    """Returns the Quadratic CVaR of the given input tensor.
+
+    .. math::
+
+        \\rho (X) = \\inf_\\omega \\left\\{\\omega + \\lambda || \\min\\{0, X + \\omega\\}||_2\\right\\}.
+
+    for :math:`\lambda\geq1`.
+
+    References:
+        - Buehler, Hans, Statistical Hedging (March 1, 2019). Available at SSRN: http://dx.doi.org/10.2139/ssrn.2913250
+          (See Conclusion.)
+
+    Args:
+        input (torch.Tensor): The input tensor.
+        lam (float): :math:`lambda`.
+        dim (int, optional): The dimension to sort along.
+
+    Returns:
+        torch.Tensor
+
+    Examples:
+        >>> from pfhedge.nn.functional import quadratic_cvar
+        >>>
+        >>> input = -torch.arange(10.0)
+        >>> input
+        tensor([-0., -1., -2., -3., -4., -5., -6., -7., -8., -9.])
+        >>> quadratic_cvar(input, 2.0)
+        tensor(7.9750)
+    """  # NOQA
+    output_target = torch.as_tensor(1 / (2 * lam))
+    if dim:
+        base = input.mean(dim=dim)
+        input -= base.unsqueeze(dim=dim)
+
+        def fn_target(_omega: Tensor) -> Tensor:
+            if not dim:
+                raise AssertionError("Unknown Error (dim is modified inappropriately.)")
+            return fn.relu(-_omega.unsqueeze(dim=dim) - input).mean(dim=dim)
+
+    else:
+        base = input.mean()
+        input -= base
+
+        def fn_target(_omega: Tensor) -> Tensor:
+            return fn.relu(-_omega - input).mean()
```

Comment:
dim == 0の扱いが怪しい。おそらく整理するとこんな感じで書ける。unsqueezeするのではなく、本当に不要になったときにmeanなりで次元を潰した方が取り回しやすいはず。

```suggestion
    output_target = torch.as_tensor(1 / (2 * lam))
    base = input.mean(dim=dim, keepdim=True)
    input = input - base

    def fn_target(omega: Tensor) -> Tensor:
        return fn.relu(-omega - input).mean(dim=dim)
```